### PR TITLE
Implements asynchronous streaming

### PR DESCRIPTION
**Pull request by elijahbenizzy**
This uses a different API than synchronous streaming. We're going to update the synchronous API to match it.

See https://github.com/DAGWorks-Inc/burr/issues/64

## Changes
- [ ] Adds asynchronous streaming API
- [ ] Adds example

## How I tested this
- [ ] Various unit tests
- [ ] Manual implementation of prod

## Notes

We're going to make a backwards-incompatible change to the old streaming API to make it much simpler. Will be easier to use.

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

**Comment by elijahbenizzy** [comment]
TODO on this PR:
- [ ] Change the synchronous API
    - [ ] Use the same set of classes for `Action`
    - [ ] Update
- [ ] Update documentation
    - [ ] Add notes on synchronous versus asynchronous (in general)
    - [ ] Update streaming
    - [ ] Add reference to the old way/migration
- [ ] Features
    - [ ] Allow streaming to return a non-state object (default)
    - [ ] Use a centralized type alias for streaming result (rather than `Union[...]`)
- [ ] Tests
     - [ ] Add more tests for error conditions
     - [ ] Test error conditions manually
- [ ] Examples
    - [ ] Polish up async streaming example

File: burr/core/action.py [hunk 1]
```diff
@@ -475,12 +482,24 @@ def is_async(self) -> bool:
         return inspect.iscoroutinefunction(self._fn)
 
 
+StreamType = Tuple[dict, Optional[State]]
```
Comment (elijahbenizzy) [inline burr/core/action.py:485]: TODO -- go over and use these when appropriate

File: burr/core/action.py [hunk 2]
```diff
@@ -584,39 +656,140 @@ def __init__(
...
+            self._result = self._process_result(result, state)
+            raise StopIteration
+        return result
 
-    def generator(self):
-        try:
-            if self._result is not None:
-                # this means we're already thorugh it, we can just call it a day
+    def __iter__(self):
+        def g():
```
Comment (elijahbenizzy) [inline burr/core/action.py:670]: g -> gen

File: burr/core/action.py [hunk 3]
```diff
@@ -674,10 +868,21 @@ def __init__(
...
+
+    def _stream_run_and_update(self, state: State, **run_kwargs) -> GeneratorReturnType:
+        yield from self._fn(state, **self._bound_params, **run_kwargs)
+
     def stream_run_and_update(
         self, state: State, **run_kwargs
-    ) -> Generator[dict, None, tuple[dict, State]]:
-        return (yield from self._fn(state, **self._bound_params, **run_kwargs))
+    ) -> Union[AsyncGeneratorReturnType, GeneratorReturnType]:
+        if self.is_async():
```
Comment (elijahbenizzy) [inline burr/core/action.py:883]: Q: can this just inherit?

File: burr/core/application.py [hunk 4]
```diff
@@ -227,24 +229,103 @@ def _run_single_step_action(
...
+        next_result = result
+        result = item
+        if next_result is not None:
+            yield next_result, None
+        # yield item, None
+    state_update = _run_reducer(action, state, result, action.name)
+    _validate_result(result, action.name)
+    _validate_reducer_writes(action, state_update, action.name)
+    yield result, state_update
+    # action.validate_inputs(inputs)
```
Comment (elijahbenizzy) [inline burr/core/application.py:319]: TODO delete the next stuff

File: burr/core/application.py [hunk 5]
```diff
@@ -947,11 +1035,219 @@ async def astream_result(
...
+                    raise ValueError("TODO -- convert")
+                generator = _arun_single_step_streaming_action(
+                    next_action, self._state, action_inputs
+                )
+                return next_action, AsyncStreamingResultContainer(
+                    generator, self._state, process_result, callback
+                )
+            else:
+                if not next_action.is_async():
+                    raise ValueError("TODO -- convert")
```
Comment (elijahbenizzy) [inline burr/core/application.py:1229]: TODO -- fix this to be a better error message

File: examples/streaming-overview/async_application.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    async for item in result:
+        print(item["response"]["content"], end="")
+    result, state = await result.get()
+    print(result["response"]["content"])
+
+
+async def main():
+    app = application()
+    await it(app)
+    await it(app)
```
Comment (elijahbenizzy) [inline examples/streaming-overview/async_application.py:198]: delete

File: burr/core/action.py [hunk 7]
```diff
@@ -584,39 +656,140 @@ def __init__(
...
-        )  # exhausts the generator, producing the `_result` variable
+        self.streaming_result_generator = streaming_result_generator
+        self._initial_state = initial_state
+        self._process_result = process_result
+        self._callback = callback
+        self._result = None
+        self._callback_realized = False
+
+    async def __anext__(self):
+        # TODO -- ensure this works
```
Comment (skrawcz) [inline burr/core/action.py:746]: what needs to be done to ensure this?

File: burr/core/action.py [hunk 8]
```diff
@@ -584,39 +656,140 @@ def __init__(
...
+
+    async def get(self) -> tuple[Optional[dict], State]:
+        # exhaust the generator
+        async for _ in self:
+            pass
+
         return self._result
 
+    @staticmethod
+    def pass_through(results: dict, final_state: State) -> "AsyncStreamingResultContainer":
```
Comment (skrawcz) [inline burr/core/action.py:782]: doc string?

File: burr/core/action.py [hunk 9]
```diff
@@ -584,39 +656,140 @@ def __init__(
...
+        if self._result is not None:
+            # we're done, and we've run through it
+            raise StopAsyncIteration
+        result, state = await self.streaming_result_generator.__anext__()
+        if state is not None:  # we're done -- we've hit the last one
+            self._result = self._process_result(result, state)
+            raise StopAsyncIteration
+        return result
+
+    def __aiter__(self):
```
Comment (skrawcz) [inline burr/core/action.py:756]: doc string?

File: burr/core/action.py (hunk 7 above)
Comment (skrawcz) [reply burr/core/action.py:746]: also doc string?

File: burr/core/application.py [hunk 10]
```diff
@@ -227,24 +229,92 @@ def _run_single_step_action(
...
     _validate_result(result, action.name)
-    new_state = _run_reducer(action, state, result, action.name)
-    return result, _state_update(state, new_state)
+    _validate_reducer_writes(action, state_update, action.name)
+    yield result, state_update
+
+
+async def _arun_multi_step_streaming_action(
+    action: AsyncStreamingAction, state: State, inputs: Optional[Dict[str, Any]]
+) -> AsyncGenerator[Tuple[dict, Optional[State]], None]:
```
Comment (skrawcz) [inline burr/core/application.py:302]: doc string

File: burr/core/application.py [hunk 11]
```diff
@@ -227,24 +229,92 @@ def _run_single_step_action(
...
     _validate_result(result, action.name)
-    _validate_reducer_writes(action, state, action.name)
-    return result, state
+    _validate_reducer_writes(action, state_update, action.name)
+    yield result, state_update
+
+
+async def _arun_single_step_streaming_action(
+    action: SingleStepStreamingAction, state: State, inputs: Optional[Dict[str, Any]]
+) -> AsyncGenerator[Tuple[dict, Optional[State]], None]:
```
Comment (skrawcz) [inline burr/core/application.py:252]: doc string just to give some context

File: examples/streaming-overview/async_application.py [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+TERMINAL_ACTIONS = [
+    "generate_image",
+    "answer_question",
+    "generate_code",
+    "prompt_for_more",
+    "unsafe_response",
+]
+
+
+async def it(app):
```
Comment (skrawcz) [inline examples/streaming-overview/async_application.py:185]: doc string on what this does

**Review by skrawcz — APPROVED** [review]
if you say it works, great!

otherwise just doc strings to help set context for the function and where /why it does what it does.

**Comment by elijahbenizzy** [comment]
Solves: https://github.com/DAGWorks-Inc/burr/issues/64

**Pull request merged**